src/flb_chunk_trace.c

```diff
@@ -0,0 +1,685 @@
+                          struct flb_input_instance *in);
+
+// To avoid double frees when enabling and disabling tracing as well
+// as avoiding race conditions when stopping fluent-bit while someone is
+// toggling tracing via the HTTP API this set of APIS with a mutex lock 
+// is used:
+//   * flb_chunk_trace_to_be_destroyed - query to see if the trace context
+//     is slated to be freed
+//   * flb_chunk_trace_set_destroy - set the trace context to be destroyed
+//     once all chunks are freed (executed in flb_chunk_trace_destroy).
+//   * flb_chunk_trace_has_chunks - see if there are still chunks using
+//     using the tracing context
+//   * flb_chunk_trace_add - increment the traces chunk count
+//   * flb_chunk_trace_sub - decrement the traces chunk count
+static inline int flb_chunk_trace_to_be_destroyed(struct flb_chunk_trace_context *ctxt)
+{
+    int ret = FLB_FALSE;
+
+    ret = (ctxt->to_destroy == 1 ? FLB_TRUE : FLB_FALSE);
+    return ret;
+}
+
+static inline int flb_chunk_trace_has_chunks(struct flb_chunk_trace_context *ctxt)
+{
+    int ret = FLB_FALSE;
+
+    ret = ((ctxt->chunks > 0) ? FLB_TRUE : FLB_FALSE);
+    return ret;
+}
+
+static inline void flb_chunk_trace_add(struct flb_chunk_trace_context *ctxt)
+{
+    ctxt->chunks++;
+}
+
+static inline void flb_chunk_trace_sub(struct flb_chunk_trace_context *ctxt)
+{
+    ctxt->chunks--;
+}
+
+static inline void flb_chunk_trace_set_destroy(struct flb_chunk_trace_context *ctxt)
+{
+    ctxt->to_destroy = 1;
+}
+
+static struct flb_output_instance *find_calyptia_output_instance(struct flb_config *config)
+{
+    struct mk_list *head = NULL;
+    struct flb_output_instance *output = NULL;
+
+    mk_list_foreach(head, &config->outputs) {
+        output = mk_list_entry(head, struct flb_output_instance, _head);
+        if (strcmp(output->p->name, "calyptia") == 0) {
+            return output;
+        }
+    }
+    return NULL;
+}
+
+static void log_cb(void *ctx, int level, const char *file, int line,
+                   const char *str)
+{
+    if (level == CIO_LOG_ERROR) {
+        flb_error("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_WARN) {
+        flb_warn("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_INFO) {
+        flb_info("[trace] %s", str);
+    }
+    else if (level == CIO_LOG_DEBUG) {
+        flb_debug("[trace] %s", str);
+    }
+}
+
+static void trace_chunk_context_destroy(struct flb_chunk_trace_context *ctxt)
+{
+    if (flb_chunk_trace_has_chunks(ctxt) == FLB_TRUE) {
+        flb_chunk_trace_set_destroy(ctxt);
+        flb_input_pause_all(ctxt->flb->config);
+        return;
+    }
+    
+    // pause all inputs, then destroy the input storage.
+    flb_input_pause_all(ctxt->flb->config);
+    // waiting for all tasks to end is key to safely stopping and destroying
+    // the fluent-bit pipeline.
+    while (flb_task_running_count(ctxt->flb->config) > 0) {
+        sleep(1);
```

Comment:
Does this sleep the pipeline?

Reply:
it does, and I am not a fan of this approach. It brings a lot of problems in the unit tests because of its non-determinism. 

The other choice is to just spike the cpu usage and simply poll rapidly until all tasks finish running. As far as I can tell there is no safe way to simply stop all running tasks on the separate pipeline I create.